hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/ContractStateMigrator.java

```diff
@@ -0,0 +1,417 @@
+            requireNonNull(key);
+            requireNonNull(value);
+            validateArgument(key.length == EVM_WORD_WIDTH_IN_INTS, "wrong length key");
+            validateArgument(value.length == EVM_WORD_WIDTH_IN_BYTES, "wrong length value");
+            if (prev != null) validateArgument(prev.length == EVM_WORD_WIDTH_IN_INTS, "wrong length prev link");
+            if (next != null) validateArgument(next.length == EVM_WORD_WIDTH_IN_INTS, "wrong length next link");
+        }
+
+        public ContractSlotLocal(@NonNull final ContractKey k, @NonNull final IterableContractValue v) {
+            this(k.getContractId(), k.getKey(), v.getValue(), v.getExplicitPrevKey(), v.getExplicitNextKey());
+        }
+    }
+
+    /** Indicates that the source or sink process processed all slots */
+    enum CompletedProcesses {
+        SOURCING,
+        SINKING
+    }
+
+    /**
+     *  Operates the source and sink processes to transform the mono-state ("from") into the modular-state ("to").
+     *  */
+    @NonNull
+    EnumSet<CompletedProcesses> transformStore(final boolean doFullValidation) {
+
+        final var slotQueue = new ArrayBlockingQueue<ContractSlotLocal>(MAXIMUM_SLOTS_IN_FLIGHT);
+
+        // Sinking and sourcing happen concurrently.  (Though sourcing is multithreaded, sinking is all on one thread.)
+        // Consider: For debugging, don't create (and start) `processSlotQueue` until after sourcing is complete. Just
+        // accumulate everything in the fromStore in the queue before sinking anything.
+
+        CompletableFuture<Void> processSlotQueue =
+                CompletableFuture.runAsync(() -> iterateOverAllQueuedSlots(slotQueue::take, doFullValidation));
+
+        final var completedTasks = EnumSet.noneOf(CompletedProcesses.class);
+
+        boolean didCompleteSourcing = iterateOverAllCurrentData(slotQueue::put, doFullValidation);
+
+        if (didCompleteSourcing) completedTasks.add(CompletedProcesses.SOURCING);
+
+        boolean didCompleteSinking = true;
+        try {
+            processSlotQueue.join();
+        } catch (CompletionException cex) {
+            final var ex = cex.getCause();
+            log.error(LOG_CAPTION + ": interrupt when sinking slots", ex);
+            didCompleteSinking = false;
+        }
+        if (didCompleteSinking) completedTasks.add(CompletedProcesses.SINKING);
+
+        return completedTasks;
+    }
+
+    /**
+     * Pull slots off the queue and add each one immediately to the modular-service state.  This is "sinking" the slots.
+     *
+     * This is single-threaded.  (But does operate concurrently with sourcing.)
+     */
+    @SuppressWarnings("java:S2142") // must rethrow InterruptedException - Sonar doesn't understand wrapping it to throw
+    void iterateOverAllQueuedSlots(
+            @NonNull final InterruptableSupplier<ContractSlotLocal> slotSource, final boolean doFullValidation) {
+        requireNonNull(slotSource);
+
+        while (true) {
```

Comment:
We should probably add an exit condition here (in addition to the SENTINEL break below), e.g. a maximum iteration size to avoid infinite loops

Reply:
hmm.

1) I'll pass in the max iteration limit as a constructor argument to `ContractStateMigrator`.  The production migrator can pull the number out of the config (max contract slots).  And the CLI can get it out of the mono contract state it holds.

2) maybe I should also `poll(timeout)` not `take` (and wait indefinitely) ... for same reason ...